fhirclient/_utils.py

```diff
@@ -0,0 +1,121 @@
+
+if TYPE_CHECKING:
+    from fhirclient.models.bundle import Bundle
+
+
+# Use forward references to avoid circular imports
+def _fetch_next_page(bundle: 'Bundle') -> Optional['Bundle']:
+    """
+    Fetch the next page of results using the `next` link provided in the bundle.
+
+    Args:
+        bundle (Bundle): The FHIR Bundle containing the `next` link.
+
+    Returns:
+        Optional[Bundle]: The next page of results as a FHIR Bundle, or None if no "next" link is found.
+    """
+    if next_link := _get_next_link(bundle):
+        sanitized_next_link = _sanitize_next_link(next_link)
+        return _execute_pagination_request(sanitized_next_link)
+    return None
+
+
+def _get_next_link(bundle: 'Bundle') -> Optional[str]:
+    """
+    Extract the `next` link from the Bundle's links.
+
+    Args:
+        bundle (Bundle): The FHIR Bundle containing pagination links.
+
+    Returns:
+        Optional[str]: The URL of the next page if available, None otherwise.
+    """
+    if not bundle.link:
+        return None
+
+    for link in bundle.link:
+        if link.relation == "next":
+            return _sanitize_next_link(link.url)
+    return None
+
+
+def _sanitize_next_link(next_link: str) -> str:
+    """
+    Sanitize the `next` link by validating its scheme and hostname against the origin server.
+
+    This function ensures the `next` link URL uses a valid scheme (`http` or `https`) and that it contains a
+    hostname. This provides a basic safeguard against malformed URLs without overly restricting flexibility.
+
+    Args:
+        next_link (str): The raw `next` link URL.
+        origin_server (str): The expected origin server hostname.
+
+    Returns:
+        str: The validated URL.
+
+    Raises:
+        ValueError: If the URL's scheme is not `http` or `https`, or if the hostname does not match the origin server.
+    """
+
+    parsed_url = urllib.parse.urlparse(next_link)
+
+    # Validate scheme and netloc (domain)
+    if parsed_url.scheme not in ["http", "https"]:
+        raise ValueError("Invalid URL scheme in `next` link.")
+    if not parsed_url.netloc:
+        raise ValueError("Invalid URL domain in `next` link.")
+
+    return next_link
+
+
+def _execute_pagination_request(sanitized_url: str) -> 'Bundle':
+    """
+    Execute the request to retrieve the next page using the sanitized URL.
+
+    Args:
+        sanitized_url (str): The sanitized URL to fetch the next page.
+
+    Returns:
+        Optional[Bundle]: The next page of results as a FHIR Bundle, or None.
+
+    Raises:
+        HTTPError: If the request fails due to network issues or server errors.
+    """
+    from fhirclient.models.bundle import Bundle  # Import here to avoid circular import
+
+    try:
+        # Use requests.get directly to make the HTTP request
+        response = requests.get(sanitized_url)
+        response.raise_for_status()
+        next_bundle_data = response.json()
+        next_bundle = Bundle(next_bundle_data)
```

Comment:
Could/should this also use `Bundle.read_from`? That would set a few more headers like `Accept: application/json` and would also send authentication headers - which... we probably need here yeah?

It would mean sending down the `FHIRServer` instance, but that should be available to the callers of this function.